test/test_nn.py

```diff
@@ -4604,6 +4605,67 @@ def test_Transformer_cell(self):
+        tgt_length = 15
+
+        # find the output of a transformer module without batch_first
+        transformer = nn.Transformer(d_model, nhead, num_encoder_layers, num_decoder_layers,
+                                     dim_feedforward, batch_first=False, dropout=dropout)
+        # use deterministic weights
+        for idx, p in enumerate(transformer.parameters()):
+            x = p.data
+            sz = x.view(-1).size(0)
+            shape = x.shape
+            x = torch.cos(torch.arange(0, sz).float().view(shape))
+            p.data.copy_(x)
+        # compare the output to a transformer module with batch_first
+        transformer_batch_first = nn.Transformer(d_model, nhead, num_encoder_layers, num_decoder_layers,
+                                                 dim_feedforward, batch_first=True, dropout=dropout)
+        # also use deterministic weights
+        for idx, p in enumerate(transformer_batch_first.parameters()):
+            x = p.data
+            sz = x.view(-1).size(0)
+            shape = x.shape
+            x = torch.cos(torch.arange(0, sz).float().view(shape))
+            p.data.copy_(x)
+
+        src = torch.randn(seq_length, bsz, d_model)
+        src_mask = transformer.generate_square_subsequent_mask(seq_length).double()
+        tgt = torch.randn(tgt_length, bsz, d_model)
+        tgt_mask = transformer.generate_square_subsequent_mask(tgt_length).double()
+        memory_mask = torch.randn(tgt_length, seq_length).double()
+        src_key_padding_mask = torch.rand(bsz, seq_length) >= 0.5
+        tgt_key_padding_mask = torch.rand(bsz, tgt_length) >= 0.5
+        memory_key_padding_mask = torch.rand(bsz, seq_length) >= 0.5
+        output = transformer(src, tgt,
+                             src_mask=src_mask,
+                             tgt_mask=tgt_mask,
+                             memory_mask=memory_mask,
+                             src_key_padding_mask=src_key_padding_mask,
+                             tgt_key_padding_mask=tgt_key_padding_mask,
+                             memory_key_padding_mask=memory_key_padding_mask)
+
+        src_batch_first = src.permute(1, 0, 2)
+        tgt_batch_first = tgt.permute(1, 0, 2)
+        output_batch_first = transformer_batch_first(src_batch_first, tgt_batch_first,
+                                                     src_mask=src_mask,
+                                                     tgt_mask=tgt_mask,
+                                                     memory_mask=memory_mask,
+                                                     src_key_padding_mask=src_key_padding_mask,
+                                                     tgt_key_padding_mask=tgt_key_padding_mask,
+                                                     memory_key_padding_mask=memory_key_padding_mask)
+        output_ref = output_batch_first.permute(1, 0, 2)
+        np.testing.assert_allclose(output.detach().numpy(), output_ref.detach().numpy(), atol=1e-5)
```

Comment:
You will need a numpy decorator if your test depends on numpy. I'd avoid the numpy dependency and `torch.testing.assert_allclose` (though some of transformer tests did use numpy)

Reply:
Yes. I might go ahead and change this and maybe also change the other transformer tests to all use `torch.testing.assert_allclose` too (if applicable).